scalding-core/src/main/scala/com/twitter/scalding/ExecutionOptimizationRules.scala

```diff
@@ -108,24 +109,346 @@ object ExecutionOptimizationRules {
+     * Executions. In this representation, we can recursively optimize.
+     *
+     * On this simpler dag, we define optimize that ensures that
+     * after there is exactly 1 WriteExecution in the head position
+     * in the HList
+     *
+     * The first type parameter Ex[x] will either be Execution[X]
+     * or when optimized WriteExecution[X], proving that we have
+     * pushed the Write into the head position of the HList
+     */
+    sealed trait FlattenedZip[+Ex[x] <: Execution[x], +A] extends Serializable {
+      import FlattenedZip._
+
+      // this is the type of the HList representation of A
+      type H <: HList
+      // this is the type of thes HList representation of Executions that when zipped give Execution[A]
+      type ExH <: HList
+
+      // here is value for the HList of Executions that have been unzipped
+      def executions: ExH
+      // this converts the above executions back into a single Execution, but still on the HList
+      // type
+      def toExecution(ex: ExH): Execution[H]
+      def fn: H => A
+
+      // convert back to the original Execution type
+      final def execution: Execution[A] =
+        this match {
+          case Single(a) => a
+          case notSingle =>
+            Mapped(toExecution(executions), fn)
+        }
+
+      def zip[B](that: FlattenedZip[Execution, B]): FlattenedZip[Execution, (A, B)]
+      def map[B](fn: A => B): FlattenedZip[Ex, B] =
+        this match {
+          case MapZip(f1, fn1) =>
+            MapZip(f1, ComposedFunctions.ComposedMapFn(fn1, fn))
+          case notMap =>
+            MapZip(notMap, fn)
+        }
+
+      // How many writes are there in this FlattenedZip
+      def writeCount: Int = {
+        def isW[A](ex: Execution[A]): Int =
+          ex match {
+            case WriteExecution(_, _, _) => 1
+            case _ => 0
+          }
+        @annotation.tailrec
+        def loop(fz: FlattenedZip[Execution, Any], acc: Int): Int =
+          fz match {
+            case Single(w) => acc + isW(w)
+            case Many(h, rest, _) => loop(rest, acc + isW(h))
+            case MapZip(of, _) => loop(of, acc)
+          }
+
+        loop(this, 0)
+      }
+
+      /**
+       * The type of this ensures that the head of the HList
+       * must be a WriteExecution. Additionally, none of
+       * the other items are, but we don't prove that in the types,
+       * instead we prove that with tests
+       */
+      def optimize: Option[FlattenedZip[WriteExecution, A]]
+    }
+
+    object FlattenedZip {
+      import HList._
+
+      /**
+       * Here is the simplest FlattenZip: just a single Execution
+       */
+      final case class Single[+Ex[x] <: Execution[x], A](ex: Ex[A]) extends FlattenedZip[Ex, A] {
+        type H = A :: HNil.type
+        type ExH = Execution[A] :: HNil.type
+        def executions = ::(ex, HNil)
+        def toExecution(ex: Execution[A] :: HNil.type): Execution[A :: HNil.type] =
```

Comment:
minor nit, but i'd use another variable name here since your shadowing the ex from the case class definition itself.

Reply:
yep, will do.